### PR TITLE
Python101

### PR DESCRIPTION
**Pull request by mattfeng**
Crash course programming: Python 101

Things to do:
  - get all the outputs right
  - finish the code walkthrough for the Classes section

### CODE REVIEW

File: _posts/2018-01-05-python101.md [hunk 1]
```diff
@@ -0,0 +1,734 @@
...
+```
+
+Why does it print "true"? Let’s take a look at it. The ``x`` evaluates as ``True``, and ``y`` evaluates as ``not False``, which is ``True``! Both are ``True``, therefore the if statement evaluates as ``True``.
+
+You can mix and match all the Boolean conditionals to fit your needs while programming!
+
+
+# Looping
+
+##``for`` loops
```
Comment (cmnord) [inline _posts/2018-01-05-python101.md]: add a space before the `!

File: _posts/2018-01-05-python101.md [hunk 2]
```diff
@@ -0,0 +1,734 @@
...
+pi = 3.14
+pi_string = str(pi)
+
+print(my_string)
+print(my_other_string)
+print(very_long_string)
+```
+Output:
+```
+<insert printed results here>
```
Comment (cmnord) [inline _posts/2018-01-05-python101.md]: don't these need to be filled in?

File: _posts/2018-01-05-python101.md [hunk 3]
```diff
@@ -0,0 +1,734 @@
...
+Am I hungry?
+If yes: make a sandwich.
+Otherwise: leave the kitchen.
+```
+
+The coding for this is also pretty straightforward. Let’s look at some examples to get you used to ``if`` statements!
+
+```python
+if 100 > 10:
+print("true")
```
Comment (cmnord) [inline _posts/2018-01-05-python101.md]: these need to be indented!

**Review by cmnord — CHANGES_REQUESTED** [review]
Looks awesome :smile: thanks for the hard work!! Take a look at the few typos I caught, etc.

File: _posts/2018-01-05-python101.md (hunk 2 above)
Comment (cmnord) [reply _posts/2018-01-05-python101.md]: oh I see the PR message says it's still todo

File: _posts/2018-01-05-python101.md [hunk 4]
```diff
@@ -0,0 +1,734 @@
...
+8. Lists
+9. Tuples
+10. Conditionals
+11. If statements
+12. Boolean operations
+13. Looping
+14. For statements
+15. While statements
+16. Functions
+17. Classes
```
Comment (michaelsilver) [inline _posts/2018-01-05-python101.md]: Can we make the whole table of contents links to the appropriate sections? [Contributing](https://github.com/starthacking/starthacking.github.io/blob/master/CONTRIBUTING.md) has a lot of examples of linking to different sections in one page.

File: _posts/2018-01-05-python101.md [hunk 5]
```diff
@@ -0,0 +1,734 @@
...
+
+# Basics
+In programming, there are 3 types of commands:
+1. Executive
+2. Conditional
+3. Iterative
+
+
+1. **Executive** means "do something". The classic way of thinking about a basic computer program is to think about a recipe. If you have a list of executive commands as your program to make a PB&J sandwich, your program would look like:
```
Comment (michaelsilver) [inline _posts/2018-01-05-python101.md]: Looks like you're going to have to fight with the markdown a little here -- the rendered version continues the numbering -- so `1. **Executive**` became 4 instead of 1

File: _posts/2018-01-05-python101.md [hunk 6]
```diff
@@ -0,0 +1,734 @@
...
+      print("Number too small")
+  else:
+    print("Sorry that you're giving up!")
+    break
+
+if guess == to_be_guessed:
+  print("Congratulations. You made it!")
+```
+
+Let’s assume that to_be_guessed has been set to some random number outside of this block of code. The first line initializes a guess variable, and from there on it’s the while loop. The loop runs as long as the guess is incorrect. Inside the loop, the first thing we do is get a guess from the user - the input function takes a string using the prompt in quotes ("New number: "). The first if/else block checks to see if the guess is valid. We’ll talk later about other, more comprehensive ways this can be done. Inside the first if block we have another if/else block to tell the user how they did if they guessed over or under — note there’s no else block to handle if they guessed correctly. What happens if they do guess correctly? We end up passing through all the ifs without printing anything but once we return to the condition check for the while loop, we find that the condition is now false. We break out of the loop and print the congratulations statement.
```
Comment (michaelsilver) [inline _posts/2018-01-05-python101.md]: > Let’s assume that to_be_guessed has been set to some random number outside of this block of code.

I think the code block is hard to follow for a beginner as is, let's just set it to a number in the code instead of writing this sentence.

Comments line by line in the code would also really help readability. Ex:

```python
guess = int(input("New number: ")) # get a guess from the user
```

File: _posts/2018-01-05-python101.md [hunk 7]
```diff
@@ -0,0 +1,734 @@
...
+    Table
+  </td>
+  <td>
+    Height <br>
+    Width <br>
+    Depth <br>
+    Material <br>
+  </td>
+  <td>
+    Change Height (if it's an <a href="https://updesk.com/">Updesk</a>)
```
Comment (michaelsilver) [inline _posts/2018-01-05-python101.md:725]: StartHacking, brought to you by, Updesk ™️ ©️ 😆

File: _posts/2018-01-05-python101.md [hunk 8]
```diff
@@ -0,0 +1,734 @@
...
+physics.set_grade("Physics homework #1", katie, 86)
+
+calculus.set_grade("Calc homework #1", john, 100)
+calculus.set_grade("Calc homework #1", katie, 98)
+calculus.set_grade("Calc quiz #1", john, 78)
+calculus.set_grade("Calc quiz #1", katie, 84)
+
+john.get_transcript()
+katie.get_transcript()
+```
```
Comment (michaelsilver) [inline _posts/2018-01-05-python101.md]: Could you break up this code a little bit? I feel like it'd be better to show one class, explain what the code is doing (for example, what's `self`?), and then show another code block showing how objects are instantiated, etc.

**Review by michaelsilver — CHANGES_REQUESTED** [review]
Hey, great python tutorial! I agreed with many of Claire's comments and left a few of my own.

Meta suggestion: we should try and help readers follow along in larger code blocks. Line by line comments, and smaller chunk of code would help.

File: _posts/2018-01-05-python101.md [hunk 9]
```diff
@@ -0,0 +1,847 @@
...
+  <ul>
+    <li><a href="#strings">Strings</a></li>
+    <li><a href="#lists">Lists</li>
+  </ul>
+</li>
+<li><a href="#dictionaries">Dictionaries</a></li>
+<li><a href="#conditionals">Conditionals</a></li>
+  <ul>
+    <li><a href="if-statements">If statements</li>
+    <li><a href="boolean-ops">Boolean operations</li>
```
Comment (michaelsilver) [inline _posts/2018-01-05-python101.md]: "if statements" and "boolean operations" table of contents links are broken

File: _posts/2018-01-05-python101.md [hunk 10]
```diff
@@ -0,0 +1,847 @@
...
+  </ul>
+</li>
+<li><a href="looping">Looping</a>
```
Comment (michaelsilver) [inline _posts/2018-01-05-python101.md]: looping link broken

File: _posts/2018-01-05-python101.md [hunk 11]
```diff
@@ -0,0 +1,847 @@
...
+my_list = [1, 2, 3, 4, 5]
+my_list[0] = 0
+print(my_list)
+```
+Output:
+```
+[0, 2, 3, 4, 5]
+```
+
+There are other sequences in Python, such as tuples and dictionaries. Feel free to explore them on your own! You’ll find them both similar and different to what we have discussed here--which is what makes coding versatile. Their unique qualities allows each type of sequence to be utilized in various ways.
```
Comment (michaelsilver) [inline _posts/2018-01-05-python101.md]: maybe add links for tuples & dictionaries (if you're not going into them here)?

File: _posts/2018-01-05-python101.md [hunk 12]
```diff
@@ -0,0 +1,847 @@
...
+elif a > b:
+  print(a)
+```
+Output:
+```
+100
+```
+
+If the ``if``/``elif`` statement is satisfied (i.e. it is true), then it will run whatever is indented below it. Indentation is essential in Python. Any indented code will only run if the code that it is under is true. You can think of code as a giant file cabinet, where each line of code is a folder, and indented lines are files inside the folder. You can only access the "indented" files if the folder it is in allows you to.
+You’ve seen if statements and you’ve seen ``elif`` statements. What exactly are ``elif`` statements? You can think of ``elif`` as short for "else if". The program only checks these statements if whatever above them is false. They must be used after an if statement, but you can also have a long line of elif statements if that is what you need. Else statements are optional, but it is good practice to use them as defaults. Whatever is indented under an else statement will only run only if every other ``if`` and ``elif`` statement above it renders as false.
```
Comment (michaelsilver) [inline _posts/2018-01-05-python101.md]: missing paragraph break here -- add a line between these two paragraphs to separate them on the page

File: _posts/2018-01-05-python101.md [hunk 13]
```diff
@@ -0,0 +1,847 @@
...
+Packing sandwich #1
+Packing sandwich #2
+Packing sandwich #3
+Packing sandwich #4
+Packing sandwich #5
+```
+
+There’s a couple things to unpack here. On the first line, we begin our ``for`` loop. We’re looping over every ``x`` in the range from 1 to 5. Why 5, you might ask, since it looks like we’re going from 1 to 6? This is because of the parameters that the range function takes: ``range([start], stop[, step])``. Whenever you use ``range``, you need to give it a stop value, and ``range`` will generate numbers up to, but not including, that value. You can optionally pass it any start value, and that is inclusive. Otherwise, ``range`` will start with 0. ``step`` specifies the difference between each number in the sequence, so you could use a ``step`` of 2 to have range generate just even numbers, for example.
+
+On the second line, we have the actual body of our ``for`` loop. This is the code that the loop will repeat for each value of x. In this simple example, it’s just a print statement. But note the indent: this is what tells Python to repeat this particular block. This is also a nice reminder of casting and string concatenation!
```
Comment (michaelsilver) [inline _posts/2018-01-05-python101.md]: i don't think you defined "casting" before you used it here.

File: _posts/2018-01-05-python101.md [hunk 14]
```diff
@@ -0,0 +1,847 @@
...
+  if guess > 0: # guess is valid only if it is positive
+    if guess > to_be_guessed: # let the user know if their guess is above/below
+      print("Number too large")
+    elif guess < to_be_guessed:
+      print("Number too small")
+  else:
+    print("Sorry that you're giving up!")
+    break
+
+if guess == to_be_guessed:  # we need this b/c its possible that they gave up
```
Comment (michaelsilver) [inline _posts/2018-01-05-python101.md:583]: is this comment out of place?

File: _posts/2018-01-05-python101.md [hunk 15]
```diff
@@ -0,0 +1,847 @@
...
+3
+4
+```
+
+Let’s break this down line by line. On the first line, we’ve created a counter variable i and set it to 1. The second line is the beginning of our loop: we’re going to keep running it until the condition ``i < 5`` is met. The next two indented lines are the block that will repeat. The first indented line is just a print statement, but the second one is more interesting: we’re incrementing the counter i by 1. i += 1 is just a more compact way of writing i = i+1, and it makes sure that we don’t endlessly get stuck in the loop: if we never increased i, the condition ``i < 5`` would always be true and we’d have an infinite loop which would never end. This is not something you usually want in your code. What about the condition? What if we changed it to i <= 5? We’d print the numbers from 1 to 5 this time, instead of 1 to 4.
+
+There are a bunch of numerical examples we could go over, but while loops get more interesting when we include user input. Let’s go over the code for a simple number guessing game:
+
+```python
+# code to get a random number -- search for the "random" module to learn more
```
Comment (michaelsilver) [inline _posts/2018-01-05-python101.md]: nit: change "search" to "search online"

File: _posts/2018-01-05-python101.md [hunk 16]
```diff
@@ -0,0 +1,847 @@
...
+import random
+to_be_guessed = random.randint(1, 100)
```
Comment (michaelsilver) [inline _posts/2018-01-05-python101.md]: couldn't hurt to include a comment saying "set to_be_guessed to a random integer between 1 and 100"

File: _posts/2018-01-05-python101.md [hunk 17]
```diff
@@ -0,0 +1,847 @@
...
+
+<dl>
+  <lt>class</lt>
+  <lt>double</lt>
+  <lt>variable</lt>
+  <lt>integer</lt>
+  <lt>float</lt>
+  <lt>object</lt>
+  <lt>instance</lt>
+</dl>
```
Comment (michaelsilver) [inline _posts/2018-01-05-python101.md]: is the glossary just supposed to be a list of words like this or are you going to hyperlink as well?

File: _posts/2018-01-05-python101.md [hunk 18]
```diff
@@ -0,0 +1,847 @@
...
+  def set_grade(self, assignment, student, grade):
+    self.assignments[assignment][student] = grade # assign a grade to a Student
+
+  def get_overall_grade_for(self, student):
+    sum_scores = 0  # need an accumulator (see Glossary) to compute average
+    num_graded = 0  # need to keep track of total number of grades for average
+    for assignment, grades in self.assignments.items(): # iterate over all items in self.assignments
+      if student in grades.keys():  # look to see if the student has a grade
+        sum_scores += grades[student]  # add that grade to the accumulator
+        num_graded += 1 # increment number of graded assignments (the denominator)
```
Comment (michaelsilver) [inline _posts/2018-01-05-python101.md:880]: if you're going do all this dictionary stuff in the example for classes, you'll need to fill in that dictionaries section above

**Review by michaelsilver — CHANGES_REQUESTED** [review]
Awesome! I left a few more small comments. Great job adding comments to the large code blocks and breaking down the classes section into chunks!

What's your plan for the Dictionaries and Glossary sections?

File: _posts/2018-01-05-python101.md (hunk 14 above)
Comment (mattfeng) [reply _posts/2018-01-05-python101.md:583]: no, the loop might have been broken because they gave up, so outside, we need to confirm that they actually get the right guess

**Comment by michaelsilver** [comment]
@mattfeng this somehow never got linked to on the frontpage. Could you look into this and make another PR?
